<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_04) on Tue Sep 27 11:17:31 EDT 2005 -->
<TITLE>
Index
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="./stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Index";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="./overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Package</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="./overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Index</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="./help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="./index.html?index-all.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="index-all.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="./allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="./allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<A HREF="#_A_">A</A> <A HREF="#_B_">B</A> <A HREF="#_C_">C</A> <A HREF="#_D_">D</A> <A HREF="#_E_">E</A> <A HREF="#_F_">F</A> <A HREF="#_G_">G</A> <A HREF="#_H_">H</A> <A HREF="#_I_">I</A> <A HREF="#_J_">J</A> <A HREF="#_L_">L</A> <A HREF="#_M_">M</A> <A HREF="#_N_">N</A> <A HREF="#_O_">O</A> <A HREF="#_P_">P</A> <A HREF="#_R_">R</A> <A HREF="#_S_">S</A> <A HREF="#_T_">T</A> <A HREF="#_U_">U</A> <A HREF="#_V_">V</A> <A HREF="#_W_">W</A> <A HREF="#_Y_">Y</A> <HR>
<A NAME="_A_"><!-- --></A><H2>
<B>A</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#A"><B>A</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/FeatureList.html#add(javagene.seq.FeatureI)"><B>add(FeatureI)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Add specified feature to the end of the list.
<DT><A HREF="./javagene/seq/FeatureList.html#add(java.util.Collection)"><B>add(Collection&lt;FeatureI&gt;)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Add all features in the specified list or collection to this list.
<DT><A HREF="./javagene/seq/AminoAcid.html#ALA"><B>ALA</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq"><B>AminoAcid</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>An amino acid object with codon/residue utility methods.<DT><A HREF="./javagene/seq/AminoAcid.html#ARG"><B>ARG</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/AminoAcid.html#ASN"><B>ASN</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/AminoAcid.html#ASP"><B>ASP</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Feature.html#attributes()"><B>attributes()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get the string of key/value attributes.
<DT><A HREF="./javagene/seq/FeatureList.html#attributeValues(java.lang.String)"><B>attributeValues(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a collection of the unique values for the specified key.
</DL>
<HR>
<A NAME="_B_"><!-- --></A><H2>
<B>B</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#B"><B>B</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>G or C or T/U (Anything but A).
<DT><A HREF="./javagene/seq/SeqStranded.html#baseSeq()"><B>baseSeq()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Get the base sequence that was specified in the constructor.
<DT><A HREF="./javagene/seq/Location.html#bioEnd()"><B>bioEnd()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Get end index, in biocoordinates.
<DT><A HREF="./javagene/seq/Location.html#bioStart()"><B>bioStart()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Get start index, in biocoordinates.
<DT><A HREF="./javagene/seq/Location.html#bioStrand()"><B>bioStrand()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Get character representation of strand.
<DT><A HREF="./javagene/seq/AminoAcid.html#blosum62(javagene.seq.AminoAcid, javagene.seq.AminoAcid)"><B>blosum62(AminoAcid, AminoAcid)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>Get BLOSUM62 distance between two AminoAcids.
<DT><A HREF="./javagene/seq/FeatureList.html#bounds()"><B>bounds()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>The union of all locations of all features in this list, mapped to the positive strand.
<DT><A HREF="./javagene/seq/SeqBig.html#bounds()"><B>bounds()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq">SeqBig</A>
<DD>Get the bounding location of this sequence.
<DT><A HREF="./javagene/seq/SeqFast.html#bounds()"><B>bounds()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>Get the bounding location of this sequence.
<DT><A HREF="./javagene/seq/SeqI.html#bounds()"><B>bounds()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>
<DD>Get the location that specifies the bounds of this sequence.
<DT><A HREF="./javagene/seq/SeqStranded.html#bounds()"><B>bounds()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Get the bounding location of this sequence.
</DL>
<HR>
<A NAME="_C_"><!-- --></A><H2>
<B>C</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#C"><B>C</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Nucleotide.html#complement(java.lang.String, javagene.seq.Nucleotide.Type)"><B>complement(String, Nucleotide.Type)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Complement a string of IUPAC nucleotides.
<DT><A HREF="./javagene/seq/Location.html#contains(javagene.seq.Location)"><B>contains(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location contains the other.
<DT><A HREF="./javagene/seq/AminoAcid.html#CYS"><B>CYS</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
</DL>
<HR>
<A NAME="_D_"><!-- --></A><H2>
<B>D</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#D"><B>D</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>A or G or T/U (Anything but C).
<DT><A HREF="./javagene/seq/SeqBig.html#description()"><B>description()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq">SeqBig</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/SeqFast.html#description()"><B>description()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/SeqI.html#description()"><B>description()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>
<DD>Get the text description of the sequence.
<DT><A HREF="./javagene/seq/SeqStranded.html#description()"><B>description()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Location.html#distance(javagene.seq.Location)"><B>distance(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return distance between this location and the other location.
<DT><A HREF="./javagene/seq/Nucleotide.html#dnaComplement(java.lang.String)"><B>dnaComplement(String)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Complement a string of IUPAC DNA nucleotides (output A C T G only).
<DT><A HREF="./javagene/seq/Nucleotide.html#dnaReverseComplement(java.lang.String)"><B>dnaReverseComplement(String)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Reverse and complement a string of IUPAC DNA nucleotides (A C T G only).
<DT><A HREF="./javagene/seq/Location.html#downstream(int)"><B>downstream(int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return the adjacent location of specified length directly downstream of this location.
</DL>
<HR>
<A NAME="_E_"><!-- --></A><H2>
<B>E</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Location.html#end()"><B>end()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Get the ending index.
<DT><A HREF="./javagene/seq/Location.html#endsAfter(javagene.seq.Location)"><B>endsAfter(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location ends after other location ends.
<DT><A HREF="./javagene/seq/Location.html#endsBefore(javagene.seq.Location)"><B>endsBefore(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location ends before other location ends.
<DT><A HREF="./javagene/seq/Location.html#equals(javagene.seq.Location)"><B>equals(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Compare locations for equality.
</DL>
<HR>
<A NAME="_F_"><!-- --></A><H2>
<B>F</B></H2>
<DL>
<DT><A HREF="./javagene/io/Fasta.html" title="class in javagene.io"><B>Fasta</B></A> - Class in <A HREF="./javagene/io/package-summary.html">javagene.io</A><DD>Read and write sequences as Fasta files.<DT><A HREF="./javagene/seq/Feature.html" title="class in javagene.seq"><B>Feature</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A Feature corresponds to a single row in a GFF file.<DT><A HREF="./javagene/seq/Feature.html#Feature(javagene.seq.Feature)"><B>Feature(Feature)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Make a copy of the specified feature.
<DT><A HREF="./javagene/seq/Feature.html#Feature(java.lang.String, java.lang.String, java.lang.String, javagene.seq.Location, java.lang.Double, int, java.lang.String)"><B>Feature(String, String, String, Location, Double, int, String)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Construct a new Feature from raw data (usually a GFF row).
<DT><A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq"><B>FeatureI</B></A> - Interface in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A feature on a sequence (for example, an exon or a gene), defined by a location
 and a set of attributes encoded as key/value pairs.<DT><A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq"><B>FeatureList</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A list of FeatureI objects implemented using a Java ArrayList; corresponds to a GFF file.<DT><A HREF="./javagene/seq/FeatureList.html#FeatureList()"><B>FeatureList()</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Construct an empty list.
<DT><A HREF="./javagene/seq/FeatureList.html#FeatureList(java.util.Collection)"><B>FeatureList(Collection&lt;FeatureI&gt;)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Construct a new list containing the same features
 as the specified list.
<DT><A HREF="./javagene/seq/Feature.html#frame()"><B>frame()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get frame (aka phase).
<DT><A HREF="./javagene/seq/Location.html#fromBio(int, int, char)"><B>fromBio(int, int, char)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Create location from "biocoordinates", as in GFF file.
<DT><A HREF="./javagene/seq/Location.html#fromBioExt(int, int, char, int)"><B>fromBioExt(int, int, char, int)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Create a location from MAF file coordinates, which represent negative
 strand locations as the distance from the end of the sequence.
</DL>
<HR>
<A NAME="_G_"><!-- --></A><H2>
<B>G</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#G"><B>G</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Feature.html#getAttribute(java.lang.String)"><B>getAttribute(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get value of specified attribute key.
<DT><A HREF="./javagene/seq/FeatureI.html#getAttribute(java.lang.String)"><B>getAttribute(String)</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>Get the attribute value for this key.
<DT><A HREF="./javagene/io/Gff.html" title="class in javagene.io"><B>Gff</B></A> - Class in <A HREF="./javagene/io/package-summary.html">javagene.io</A><DD>Read and write FeatureLists as GFF/GTF formatted files.<DT><A HREF="./javagene/seq/AminoAcid.html#GLN"><B>GLN</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/AminoAcid.html#GLU"><B>GLU</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/AminoAcid.html#GLY"><B>GLY</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Feature.html#group()"><B>group()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get the first item (everything before first semicolon, if it has one)
 in the attribute field, which is assumed to
 be a group identifer.
<DT><A HREF="./javagene/seq/FeatureI.html#group()"><B>group()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>Get the group id of the feature.
<DT><A HREF="./javagene/seq/FeatureList.html#groupValues()"><B>groupValues()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a collection of all unique group ids in the list, as defined
 by the group() method of the features.
</DL>
<HR>
<A NAME="_H_"><!-- --></A><H2>
<B>H</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#H"><B>H</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>A or C or T/U (Anything but G).
<DT><A HREF="./javagene/seq/Feature.html#hasAttribute(java.lang.String)"><B>hasAttribute(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Feature.html#hasAttribute(java.lang.String, java.lang.String)"><B>hasAttribute(String, String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/FeatureI.html#hasAttribute(java.lang.String)"><B>hasAttribute(String)</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>Check if the feature has a value defined for the specified key.
<DT><A HREF="./javagene/seq/FeatureI.html#hasAttribute(java.lang.String, java.lang.String)"><B>hasAttribute(String, String)</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>Check if the feature attributes include the specified key/value pair.
<DT><A HREF="./javagene/seq/FeatureList.html#hasAttribute(java.lang.String)"><B>hasAttribute(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Check if any feature in list has the specified attribute key.
<DT><A HREF="./javagene/seq/FeatureList.html#hasAttribute(java.lang.String, java.lang.String)"><B>hasAttribute(String, String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Check if any feature in list has the specified attribute key/value pair.
<DT><A HREF="./javagene/seq/FeatureList.html#hasGaps(int)"><B>hasGaps(int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Check size of gaps between successive features in list.
<DT><A HREF="./javagene/seq/LocIterator.html#hasNext(int, int)"><B>hasNext(int, int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq">LocIterator</A>
<DD>Check if next window of specified size is available.
<DT><A HREF="./javagene/seq/LocIterator.html#hasNext()"><B>hasNext()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq">LocIterator</A>
<DD>Check if next window of default size is available.
<DT><A HREF="./javagene/seq/AminoAcid.html#HIS"><B>HIS</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
</DL>
<HR>
<A NAME="_I_"><!-- --></A><H2>
<B>I</B></H2>
<DL>
<DT><A HREF="./javagene/seq/SeqBig.html#id()"><B>id()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq">SeqBig</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/SeqFast.html#id()"><B>id()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/SeqI.html#id()"><B>id()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>
<DD>Get the (single-word) identifier of the sequence.
<DT><A HREF="./javagene/seq/SeqStranded.html#id()"><B>id()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/AminoAcid.html#ILE"><B>ILE</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Location.html#intersection(javagene.seq.Location)"><B>intersection(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return the intersection, or null if no overlap.
<DT><A HREF="./javagene/seq/Location.html#isAfter(javagene.seq.Location)"><B>isAfter(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location is entirely after the other location (no overlap).
<DT><A HREF="./javagene/seq/Location.html#isBefore(javagene.seq.Location)"><B>isBefore(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location is entirely before other location (no overlap).
<DT><A HREF="./javagene/seq/Nucleotide.html#isComplement(char, char)"><B>isComplement(char, char)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Check if two symbols complement one another.
<DT><A HREF="./javagene/seq/Location.html#isNegative()"><B>isNegative()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if location is on negative strand.
<DT><A HREF="./javagene/seq/Location.html#isSameStrand(javagene.seq.Location)"><B>isSameStrand(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location is on same strand as other location.
<DT><A HREF="./javagene/seq/AminoAcid.html#isSynonym(java.lang.String)"><B>isSynonym(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>Check if a three-letter Dna/Rna codon codes for this AminoAcid.
<DT><A HREF="./javagene/seq/Nucleotide.html#isWobbleComplement(char, char)"><B>isWobbleComplement(char, char)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Check if two symbols complement one another, accepting wobble pairs (G-U, G-T) in addition to 
 standard (A-T, A-U, G-C) pairs.
<DT><A HREF="./javagene/seq/Location.html#iterator()"><B>iterator()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Create a location iterator over this location with a window size of 1 and
 an increment of +1 (successive symbols from start to end).
<DT><A HREF="./javagene/seq/Location.html#iterator(int, int)"><B>iterator(int, int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Create a location iterator over this location, 
 using specified window size and increment.
</DL>
<HR>
<A NAME="_J_"><!-- --></A><H2>
<B>J</B></H2>
<DL>
<DT><A HREF="./javagene/io/package-summary.html"><B>javagene.io</B></A> - package javagene.io<DD>Reading and writing files.<DT><A HREF="./javagene/seq/package-summary.html"><B>javagene.seq</B></A> - package javagene.seq<DD>Sequences, features on sequences, and the locations of features.<DT><A HREF="./javagene/util/package-summary.html"><B>javagene.util</B></A> - package javagene.util<DD>Housekeeping.</DL>
<HR>
<A NAME="_L_"><!-- --></A><H2>
<B>L</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Location.html#length()"><B>length()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Get length of range.
<DT><A HREF="./javagene/seq/AminoAcid.html#LEU"><B>LEU</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Feature.html#location()"><B>location()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get location of feature.
<DT><A HREF="./javagene/seq/FeatureI.html#location()"><B>location()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>Get the location of the feature.
<DT><A HREF="./javagene/seq/Location.html" title="class in javagene.seq"><B>Location</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A location on a sequence.<DT><A HREF="./javagene/seq/Location.html#Location(int, int)"><B>Location(int, int)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Construct new location from coordinates.
<DT><A HREF="./javagene/seq/Location.html#Location(javagene.seq.Location)"><B>Location(Location)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Clone other location.
<DT><A HREF="./javagene/seq/SeqFast.html#locationOf(java.lang.String)"><B>locationOf(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>Return the location of the parameter string in this sequence.
<DT><A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq"><B>LocIterator</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>Move a sliding window over a Location.<DT><A HREF="./javagene/seq/LocIterator.html#LocIterator(javagene.seq.Location, int, int)"><B>LocIterator(Location, int, int)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq">LocIterator</A>
<DD>Construct an iterator that slides a window over a Location.
<DT><A HREF="./javagene/util/Log.html" title="class in javagene.util"><B>Log</B></A> - Class in <A HREF="./javagene/util/package-summary.html">javagene.util</A><DD>Write text messages to the console.<DT><A HREF="./javagene/util/Log.html#log()"><B>log()</B></A> - 
Static method in class javagene.util.<A HREF="./javagene/util/Log.html" title="class in javagene.util">Log</A>
<DD>Write a blank line.
<DT><A HREF="./javagene/util/Log.html#log(java.lang.String)"><B>log(String)</B></A> - 
Static method in class javagene.util.<A HREF="./javagene/util/Log.html" title="class in javagene.util">Log</A>
<DD>Write a line to the console.
<DT><A HREF="./javagene/seq/AminoAcid.html#lookup(java.lang.String)"><B>lookup(String)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>Find the AminoAcid object corresponding to a three-letter Dna/Rna codon.
<DT><A HREF="./javagene/seq/Nucleotide.html#lookup(java.lang.String)"><B>lookup(String)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Find the Nucleotide object corresponding to an IUPAC symbol.
<DT><A HREF="./javagene/seq/AminoAcid.html#LYS"><B>LYS</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
</DL>
<HR>
<A NAME="_M_"><!-- --></A><H2>
<B>M</B></H2>
<DL>
<DT><A HREF="./javagene/seq/SeqStranded.html#mapFromBase(javagene.seq.Location)"><B>mapFromBase(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Map a location on the base sequence to the equivalent
 genomic location.
<DT><A HREF="./javagene/seq/SeqStranded.html#mapToBase(javagene.seq.Location)"><B>mapToBase(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Map a genomic location to the equivalent location
 on the base sequence.
<DT><A HREF="./javagene/seq/Nucleotide.html#matches(javagene.seq.Nucleotide)"><B>matches(Nucleotide)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Check if a Nucleotide object matches this object, allowing ambiguous
 matches.
<DT><A HREF="./javagene/seq/Nucleotide.html#matches(java.lang.String)"><B>matches(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Check if a IUPAC nucleotide symbol matches this object, allowing ambiguous matches.
<DT><A HREF="./javagene/seq/AminoAcid.html#MET"><B>MET</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Location.html#minus()"><B>minus()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return location that is in same position on negative strand.
</DL>
<HR>
<A NAME="_N_"><!-- --></A><H2>
<B>N</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#N"><B>N</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Any nucleotide.
<DT><A HREF="./javagene/seq/LocIterator.html#next()"><B>next()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq">LocIterator</A>
<DD>Get next window of default size, then increment position by default amount.
<DT><A HREF="./javagene/seq/LocIterator.html#next(int, int)"><B>next(int, int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq">LocIterator</A>
<DD>Get next window of specified size, then increment position by specified amount.
<DT><A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq"><B>Nucleotide</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A nucleotide object, with Dna/Rna utility methods.<DT><A HREF="./javagene/seq/Nucleotide.Type.html" title="enum in javagene.seq"><B>Nucleotide.Type</B></A> - Enum in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>Sequence type: DNA or RNA.</DL>
<HR>
<A NAME="_O_"><!-- --></A><H2>
<B>O</B></H2>
<DL>
<DT><A HREF="./javagene/seq/FeatureList.html#omitOverlapping(java.lang.String, javagene.seq.Location, boolean)"><B>omitOverlapping(String, Location, boolean)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a list of all features that do not overlap the specified location on the specified sequence.
<DT><A HREF="./javagene/seq/Location.html#opposite()"><B>opposite()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return location that is in same position on opposite strand.
<DT><A HREF="./javagene/seq/Location.html#overlaps(javagene.seq.Location)"><B>overlaps(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location and other location overlap.
</DL>
<HR>
<A NAME="_P_"><!-- --></A><H2>
<B>P</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Location.html#percentOverlap(javagene.seq.Location)"><B>percentOverlap(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return percent overlap of two locations.
<DT><A HREF="./javagene/seq/AminoAcid.html#PHE"><B>PHE</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Location.html#plus()"><B>plus()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return location that is in same position on plus strand.
<DT><A HREF="./javagene/seq/Location.html#prefix(int)"><B>prefix(int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>The part of this location before the specified position.
<DT><A HREF="./javagene/seq/Location.html#prefix(javagene.seq.Location)"><B>prefix(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>The part of this location before the other location (not inclusive).
<DT><A HREF="./javagene/seq/AminoAcid.html#PRO"><B>PRO</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
</DL>
<HR>
<A NAME="_R_"><!-- --></A><H2>
<B>R</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#R"><B>R</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>A or G.
<DT><A HREF="./javagene/io/Fasta.html#read(java.lang.String, int)"><B>read(String, int)</B></A> - 
Static method in class javagene.io.<A HREF="./javagene/io/Fasta.html" title="class in javagene.io">Fasta</A>
<DD>Read a file containing a single sequence.
<DT><A HREF="./javagene/io/Gff.html#read(java.lang.String)"><B>read(String)</B></A> - 
Static method in class javagene.io.<A HREF="./javagene/io/Gff.html" title="class in javagene.io">Gff</A>
<DD>Read a file into a FeatureList.
<DT><A HREF="./javagene/io/Fasta.html#readBig(java.lang.String)"><B>readBig(String)</B></A> - 
Static method in class javagene.io.<A HREF="./javagene/io/Fasta.html" title="class in javagene.io">Fasta</A>
<DD>Read a Fasta file containing a single sequence into a SeqBig object.
<DT><A HREF="./javagene/io/Fasta.html#readMany(java.lang.String, int)"><B>readMany(String, int)</B></A> - 
Static method in class javagene.io.<A HREF="./javagene/io/Fasta.html" title="class in javagene.io">Fasta</A>
<DD>Read a file containing (possibly) multiple sequences.
<DT><A HREF="./javagene/seq/LocIterator.html#remainder()"><B>remainder()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq">LocIterator</A>
<DD>Get portion of bounding location that has not yet been retrieved by next() method.
<DT><A HREF="./javagene/seq/LocIterator.html#remove()"><B>remove()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq">LocIterator</A>
<DD>Unsupported.
<DT><A HREF="./javagene/seq/Nucleotide.html#reverseComplement(java.lang.String, javagene.seq.Nucleotide.Type)"><B>reverseComplement(String, Nucleotide.Type)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Reverse and complement a string of IUPAC DNA nucleotides.
<DT><A HREF="./javagene/seq/Nucleotide.html#rnaComplement(java.lang.String)"><B>rnaComplement(String)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Complement a string of IUPAC RNA nucleotides (output A C U G only).
<DT><A HREF="./javagene/seq/Nucleotide.html#rnaReverseComplement(java.lang.String)"><B>rnaReverseComplement(String)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Reverse and complement a string of IUPAC RNA nucleotides (A C U G only).
</DL>
<HR>
<A NAME="_S_"><!-- --></A><H2>
<B>S</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Feature.html#score()"><B>score()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get score.
<DT><A HREF="./javagene/seq/FeatureList.html#selectByAttribute(java.lang.String, java.lang.String)"><B>selectByAttribute(String, String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a list of all features that include the specified attribute key/value pair.
<DT><A HREF="./javagene/seq/FeatureList.html#selectByAttribute(java.lang.String)"><B>selectByAttribute(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a list of all features that include the specified attribute key.
<DT><A HREF="./javagene/seq/FeatureList.html#selectByGroup(java.lang.String)"><B>selectByGroup(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a list of all features that have the specified group id, as defined by
 the group() method of the features.
<DT><A HREF="./javagene/seq/FeatureList.html#selectByType(java.lang.String)"><B>selectByType(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a list of all features that are of the specified type, as defined by
 the type() method of the features.
<DT><A HREF="./javagene/seq/FeatureList.html#selectByUserData(java.lang.String, java.lang.Object)"><B>selectByUserData(String, Object)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a list of all features that include the specified key/value pair in their userMap().
<DT><A HREF="./javagene/seq/FeatureList.html#selectByUserData(java.lang.String)"><B>selectByUserData(String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a list of all features that include the specified key in their userMap().
<DT><A HREF="./javagene/seq/FeatureList.html#selectOverlapping(java.lang.String, javagene.seq.Location, boolean)"><B>selectOverlapping(String, Location, boolean)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a list of all features that overlap the specified location on the specified
 sequence.
<DT><A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq"><B>SeqBig</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A SeqI implementation capable of handling large sequences that do
 not fit into memory.<DT><A HREF="./javagene/seq/SeqBig.html#SeqBig(java.lang.String)"><B>SeqBig(String)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq">SeqBig</A>
<DD>Construct SeqBig object given Fasta filename.
<DT><A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq"><B>SeqFast</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A SeqI implementation that keeps the entire sequence in memory.<DT><A HREF="./javagene/seq/SeqFast.html#SeqFast(java.lang.String, java.lang.String, java.lang.String)"><B>SeqFast(String, String, String)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>Construct SeqFast object from string data.
<DT><A HREF="./javagene/seq/SeqFast.html#SeqFast(javagene.seq.SeqI)"><B>SeqFast(SeqI)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>Construct SeqFast object from an arbitrary SeqI object.
<DT><A HREF="./javagene/seq/SeqI.html" title="interface in javagene.seq"><B>SeqI</B></A> - Interface in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A sequence of either nucleotides or amino acids.<DT><A HREF="./javagene/seq/Feature.html#seqname()"><B>seqname()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get the sequence name.
<DT><A HREF="./javagene/seq/FeatureI.html#seqname()"><B>seqname()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>Get the sequence name.
<DT><A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq"><B>SeqStranded</B></A> - Class in <A HREF="./javagene/seq/package-summary.html">javagene.seq</A><DD>A SeqI wrapper that presents a stranded coordinate system 
 mapped to an underlying SeqI object.<DT><A HREF="./javagene/seq/SeqStranded.html#SeqStranded(javagene.seq.SeqI, javagene.seq.Nucleotide.Type)"><B>SeqStranded(SeqI, Nucleotide.Type)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Construct from an arbitrary SeqI object.
<DT><A HREF="./javagene/seq/SeqStranded.html#SeqStranded(javagene.seq.SeqStranded)"><B>SeqStranded(SeqStranded)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Construct a clone.
<DT><A HREF="./javagene/seq/SeqStranded.html#SeqStranded(javagene.seq.SeqI, javagene.seq.Location, javagene.seq.Nucleotide.Type)"><B>SeqStranded(SeqI, Location, Nucleotide.Type)</B></A> - 
Constructor for class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Construct from SeqI using specified location as the bounding
 location.
<DT><A HREF="./javagene/seq/AminoAcid.html#SER"><B>SER</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/FeatureList.html#sortByStart()"><B>sortByStart()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Create a new list that is ordered by the starting index of the features' locations.
<DT><A HREF="./javagene/seq/Feature.html#source()"><B>source()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get source (aka method).
<DT><A HREF="./javagene/seq/FeatureList.html#splice(javagene.seq.SeqI)"><B>splice(SeqI)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Concatenate successive portions of the specified sequence
 using the feature locations in the list.
<DT><A HREF="./javagene/seq/Location.html#start()"><B>start()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Get starting index (origin 0).
<DT><A HREF="./javagene/seq/Location.html#startsAfter(javagene.seq.Location)"><B>startsAfter(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location starts after the other location starts.
<DT><A HREF="./javagene/seq/Location.html#startsBefore(javagene.seq.Location)"><B>startsBefore(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Check if this location starts before other location starts.
<DT><A HREF="./javagene/seq/AminoAcid.html#STOP"><B>STOP</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/SeqBig.html#subseq(javagene.seq.Location)"><B>subseq(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq">SeqBig</A>
<DD>Create a new sequence containing only the portion specified by the location parameter.
<DT><A HREF="./javagene/seq/SeqBig.html#subseq(javagene.seq.Location, java.lang.String, java.lang.String)"><B>subseq(Location, String, String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq">SeqBig</A>
<DD>Create a new sequence containing only the portion specified by the location parameter,
 with the specified id and description.
<DT><A HREF="./javagene/seq/SeqFast.html#subseq(javagene.seq.Location)"><B>subseq(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>Create a new sequence containing only the portion specified by the location parameter.
<DT><A HREF="./javagene/seq/SeqFast.html#subseq(javagene.seq.Location, java.lang.String, java.lang.String)"><B>subseq(Location, String, String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>Create a new sequence containing only the portion specified by the location parameter,
 with the specified id and description.
<DT><A HREF="./javagene/seq/SeqI.html#subseq(javagene.seq.Location)"><B>subseq(Location)</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>
<DD>Create a new sequence containing only the portion specified by the location parameter.
<DT><A HREF="./javagene/seq/SeqI.html#subseq(javagene.seq.Location, java.lang.String, java.lang.String)"><B>subseq(Location, String, String)</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>
<DD>Create a new sequence containing only the portion specified by the location parameter,
 with the specified id and description.
<DT><A HREF="./javagene/seq/SeqStranded.html#subseq(javagene.seq.Location)"><B>subseq(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Create a new sequence containing only the portion specified by the location parameter.
<DT><A HREF="./javagene/seq/SeqStranded.html#subseq(javagene.seq.Location, java.lang.String, java.lang.String)"><B>subseq(Location, String, String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Create a new sequence containing only the portion specified by the location parameter,
 with the specified id and description.
<DT><A HREF="./javagene/seq/Location.html#suffix(int)"><B>suffix(int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>The part of this location after the specified position.
<DT><A HREF="./javagene/seq/Location.html#suffix(javagene.seq.Location)"><B>suffix(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>The part of this location after the other location (not inclusive).
</DL>
<HR>
<A NAME="_T_"><!-- --></A><H2>
<B>T</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#T"><B>T</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/AminoAcid.html#THR"><B>THR</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/SeqFast.html#toReverseString(javagene.seq.Location)"><B>toReverseString(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>Get the IUPAC character string representation of the subsequence specified by 
 the location parameter, in reverse order.
<DT><A HREF="./javagene/seq/AminoAcid.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>Get the IUPAC single letter representation of this AminoAcid.
<DT><A HREF="./javagene/seq/Feature.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/FeatureI.html#toString()"><B>toString()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>A string representation of the feature.
<DT><A HREF="./javagene/seq/FeatureList.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/FeatureList.html" title="class in javagene.seq">FeatureList</A>
<DD>Return a string representation of all features in this list.
<DT><A HREF="./javagene/seq/Location.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return a string representation of location.
<DT><A HREF="./javagene/seq/LocIterator.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/LocIterator.html" title="class in javagene.seq">LocIterator</A>
<DD>Get string representation of iterator.
<DT><A HREF="./javagene/seq/Nucleotide.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>Get IUPAC representation.
<DT><A HREF="./javagene/seq/SeqBig.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq">SeqBig</A>
<DD>Get the IUPAC character string representation of this entire sequence.
<DT><A HREF="./javagene/seq/SeqBig.html#toString(javagene.seq.Location)"><B>toString(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqBig.html" title="class in javagene.seq">SeqBig</A>
<DD>Get the IUPAC character string representation of the subsequence specified by 
 the location parameter.
<DT><A HREF="./javagene/seq/SeqFast.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/SeqFast.html#toString(javagene.seq.Location)"><B>toString(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/SeqI.html#toString(javagene.seq.Location)"><B>toString(Location)</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>
<DD>Get the IUPAC character string representation of the subsequence specified by 
 the location parameter.
<DT><A HREF="./javagene/seq/SeqI.html#toString()"><B>toString()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>
<DD>Get the IUPAC character string representation of this entire sequence.
<DT><A HREF="./javagene/seq/SeqStranded.html#toString(javagene.seq.Location)"><B>toString(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Get the IUPAC character string representation of the subsequence specified by 
 the location parameter.
<DT><A HREF="./javagene/seq/SeqStranded.html#toString()"><B>toString()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>
<DD>Get the IUPAC character string representation of this entire sequence,
 as defined by the bounding location.
<DT><A HREF="./javagene/seq/SeqFast.html#toTaggedString(javagene.seq.Location, java.lang.String, java.lang.String)"><B>toTaggedString(Location, String, String)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/SeqFast.html" title="class in javagene.seq">SeqFast</A>
<DD>Get the IUPAC representation of the entire sequence, bracketing the portion
 specified by the location parameter with the specified begin and end tags.
<DT><A HREF="./javagene/seq/AminoAcid.html#translate(java.lang.String)"><B>translate(String)</B></A> - 
Static method in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>Translate from IUPAC nucleotide string to IUPAC protein string.
<DT><A HREF="./javagene/seq/AminoAcid.html#TRP"><B>TRP</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Feature.html#type()"><B>type()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get feature type, such as "exon" or "CDS".
<DT><A HREF="./javagene/seq/FeatureI.html#type()"><B>type()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>Get the feature type, for example, "exon", "CDS", etc.
<DT><A HREF="./javagene/seq/AminoAcid.html#TYR"><B>TYR</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
</DL>
<HR>
<A NAME="_U_"><!-- --></A><H2>
<B>U</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#U"><B>U</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/AminoAcid.html#UNDEFINED"><B>UNDEFINED</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Location.html#union(javagene.seq.Location)"><B>union(Location)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return the union.
<DT><A HREF="./javagene/seq/Location.html#upstream(int)"><B>upstream(int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Return the adjacent location of specified length directly upstream of this location.
<DT><A HREF="./javagene/seq/Feature.html#userData()"><B>userData()</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Feature.html" title="class in javagene.seq">Feature</A>
<DD>Get HashMap of user data.
<DT><A HREF="./javagene/seq/FeatureI.html#userData()"><B>userData()</B></A> - 
Method in interface javagene.seq.<A HREF="./javagene/seq/FeatureI.html" title="interface in javagene.seq">FeatureI</A>
<DD>Get HashMap of user data.
</DL>
<HR>
<A NAME="_V_"><!-- --></A><H2>
<B>V</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#V"><B>V</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>A or C or G (Anything but T/U).
<DT><A HREF="./javagene/seq/AminoAcid.html#VAL"><B>VAL</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/AminoAcid.html" title="class in javagene.seq">AminoAcid</A>
<DD>&nbsp;
<DT><A HREF="./javagene/seq/Nucleotide.Type.html#valueOf(java.lang.String)"><B>valueOf(String)</B></A> - 
Static method in enum javagene.seq.<A HREF="./javagene/seq/Nucleotide.Type.html" title="enum in javagene.seq">Nucleotide.Type</A>
<DD>Returns the enum constant of this type with the specified name.
<DT><A HREF="./javagene/seq/Nucleotide.Type.html#values()"><B>values()</B></A> - 
Static method in enum javagene.seq.<A HREF="./javagene/seq/Nucleotide.Type.html" title="enum in javagene.seq">Nucleotide.Type</A>
<DD>Returns an array containing the constants of this enum type, in
the order they're declared.
</DL>
<HR>
<A NAME="_W_"><!-- --></A><H2>
<B>W</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Location.html#window(int, int)"><B>window(int, int)</B></A> - 
Method in class javagene.seq.<A HREF="./javagene/seq/Location.html" title="class in javagene.seq">Location</A>
<DD>Enable a "sliding window" iteration over a location 
 to use with Java's "for" loop construct.
<DT><A HREF="./javagene/io/Fasta.html#write(java.util.ArrayList, java.lang.String)"><B>write(ArrayList&lt;SeqI&gt;, String)</B></A> - 
Static method in class javagene.io.<A HREF="./javagene/io/Fasta.html" title="class in javagene.io">Fasta</A>
<DD>Write an ArrayList of sequences to file.
<DT><A HREF="./javagene/io/Fasta.html#write(javagene.seq.SeqI, java.lang.String)"><B>write(SeqI, String)</B></A> - 
Static method in class javagene.io.<A HREF="./javagene/io/Fasta.html" title="class in javagene.io">Fasta</A>
<DD>Write a sequence to file.
<DT><A HREF="./javagene/io/Gff.html#write(javagene.seq.FeatureList, java.lang.String)"><B>write(FeatureList, String)</B></A> - 
Static method in class javagene.io.<A HREF="./javagene/io/Gff.html" title="class in javagene.io">Gff</A>
<DD>Write features in FeatureList to file.
</DL>
<HR>
<A NAME="_Y_"><!-- --></A><H2>
<B>Y</B></H2>
<DL>
<DT><A HREF="./javagene/seq/Nucleotide.html#Y"><B>Y</B></A> - 
Static variable in class javagene.seq.<A HREF="./javagene/seq/Nucleotide.html" title="class in javagene.seq">Nucleotide</A>
<DD>C or T/U.
</DL>
<HR>
<A HREF="#_A_">A</A> <A HREF="#_B_">B</A> <A HREF="#_C_">C</A> <A HREF="#_D_">D</A> <A HREF="#_E_">E</A> <A HREF="#_F_">F</A> <A HREF="#_G_">G</A> <A HREF="#_H_">H</A> <A HREF="#_I_">I</A> <A HREF="#_J_">J</A> <A HREF="#_L_">L</A> <A HREF="#_M_">M</A> <A HREF="#_N_">N</A> <A HREF="#_O_">O</A> <A HREF="#_P_">P</A> <A HREF="#_R_">R</A> <A HREF="#_S_">S</A> <A HREF="#_T_">T</A> <A HREF="#_U_">U</A> <A HREF="#_V_">V</A> <A HREF="#_W_">W</A> <A HREF="#_Y_">Y</A> 

<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="./overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Package</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="./overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Index</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="./help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="./index.html?index-all.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="index-all.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="./allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="./allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
