<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_04) on Tue Sep 27 11:17:30 EDT 2005 -->
<TITLE>
SeqStranded
</TITLE>

<META NAME="keywords" CONTENT="javagene.seq.SeqStranded class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="SeqStranded";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq"><B>PREV CLASS</B></A>&nbsp;
&nbsp;NEXT CLASS</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?javagene/seq/SeqStranded.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SeqStranded.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
javagene.seq</FONT>
<BR>
Class SeqStranded</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>javagene.seq.SeqStranded</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>SeqStranded</B><DT>extends java.lang.Object<DT>implements <A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></DL>
</PRE>

<P>
A SeqI wrapper that presents a stranded coordinate system 
 mapped to an underlying SeqI object.
 <br><br>
 The bounds of a SeqStranded object can be set to an arbitrary location; 
 they do not necessarily start at 0, and they may be on the reverse
 strand.
 <br><br>
 In JavaGene's coordinate system a numerical increase
 is always upstream. The direction (strand) of the underlying sequence
 is indicated by the strand of the bounding location.
<br><br>
 The sequence type (RNA or DNA) is always set in constructor.
<P>

<P>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#SeqStranded(javagene.seq.SeqI, javagene.seq.Location, javagene.seq.Nucleotide.Type)">SeqStranded</A></B>(<A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>&nbsp;baseSequence,
            <A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;location,
            <A HREF="../../javagene/seq/Nucleotide.Type.html" title="enum in javagene.seq">Nucleotide.Type</A>&nbsp;type)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Construct from SeqI using specified location as the bounding
 location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#SeqStranded(javagene.seq.SeqI, javagene.seq.Nucleotide.Type)">SeqStranded</A></B>(<A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>&nbsp;baseSequence,
            <A HREF="../../javagene/seq/Nucleotide.Type.html" title="enum in javagene.seq">Nucleotide.Type</A>&nbsp;type)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Construct from an arbitrary SeqI object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#SeqStranded(javagene.seq.SeqStranded)">SeqStranded</A></B>(<A HREF="../../javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>&nbsp;baseSequence)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Construct a clone.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#baseSeq()">baseSeq</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the base sequence that was specified in the constructor.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#bounds()">bounds</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the bounding location of this sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#description()">description</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the text description of the sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#id()">id</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the (single-word) identifier of the sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#mapFromBase(javagene.seq.Location)">mapFromBase</A></B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;baseLocation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Map a location on the base sequence to the equivalent
 genomic location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#mapToBase(javagene.seq.Location)">mapToBase</A></B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;genomicLocation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Map a genomic location to the equivalent location
 on the base sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#subseq(javagene.seq.Location)">subseq</A></B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;location)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new sequence containing only the portion specified by the location parameter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#subseq(javagene.seq.Location, java.lang.String, java.lang.String)">subseq</A></B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;location,
       java.lang.String&nbsp;id,
       java.lang.String&nbsp;description)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new sequence containing only the portion specified by the location parameter,
 with the specified id and description.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the IUPAC character string representation of this entire sequence,
 as defined by the bounding location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../javagene/seq/SeqStranded.html#toString(javagene.seq.Location)">toString</A></B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;location)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the IUPAC character string representation of the subsequence specified by 
 the location parameter.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="SeqStranded(javagene.seq.SeqI, javagene.seq.Location, javagene.seq.Nucleotide.Type)"><!-- --></A><H3>
SeqStranded</H3>
<PRE>
public <B>SeqStranded</B>(<A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>&nbsp;baseSequence,
                   <A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;location,
                   <A HREF="../../javagene/seq/Nucleotide.Type.html" title="enum in javagene.seq">Nucleotide.Type</A>&nbsp;type)</PRE>
<DL>
<DD>Construct from SeqI using specified location as the bounding
 location.
 Shares the underlying data.
 The start index of the bounding location may be non-zero.
 The bounding location may be on either strand.
 Note: the length of the sequence must equal
 the length of the location.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>baseSequence</CODE> - A SeqI object.<DD><CODE>location</CODE> - The bounding location.<DD><CODE>type</CODE> - Nucleotide.Type.DNA or Nucleotide.Type.RNA
<DT><B>Throws:</B>
<DD><CODE>java.lang.IllegalArgumentException</CODE> - The sequence length and location length
 were not equal.</DL>
</DL>
<HR>

<A NAME="SeqStranded(javagene.seq.SeqI, javagene.seq.Nucleotide.Type)"><!-- --></A><H3>
SeqStranded</H3>
<PRE>
public <B>SeqStranded</B>(<A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A>&nbsp;baseSequence,
                   <A HREF="../../javagene/seq/Nucleotide.Type.html" title="enum in javagene.seq">Nucleotide.Type</A>&nbsp;type)</PRE>
<DL>
<DD>Construct from an arbitrary SeqI object. Shares the underlying data.
 The bounding location starts at 0.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>baseSequence</CODE> - A SeqI object.<DD><CODE>type</CODE> - Nucleotide.Type.DNA or Nucleotide.Type.RNA.</DL>
</DL>
<HR>

<A NAME="SeqStranded(javagene.seq.SeqStranded)"><!-- --></A><H3>
SeqStranded</H3>
<PRE>
public <B>SeqStranded</B>(<A HREF="../../javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A>&nbsp;baseSequence)</PRE>
<DL>
<DD>Construct a clone. Shares the underlying data.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>baseSequence</CODE> - A SeqI object.</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="baseSeq()"><!-- --></A><H3>
baseSeq</H3>
<PRE>
public <A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A> <B>baseSeq</B>()</PRE>
<DL>
<DD>Get the base sequence that was specified in the constructor.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>The base sequence.</DL>
</DD>
</DL>
<HR>

<A NAME="bounds()"><!-- --></A><H3>
bounds</H3>
<PRE>
public <A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A> <B>bounds</B>()</PRE>
<DL>
<DD>Get the bounding location of this sequence.
 The location can start at any index, on either
 strand.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../javagene/seq/SeqI.html#bounds()">bounds</A></CODE> in interface <CODE><A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>The bounding location.</DL>
</DD>
</DL>
<HR>

<A NAME="description()"><!-- --></A><H3>
description</H3>
<PRE>
public java.lang.String <B>description</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../javagene/seq/SeqI.html#description()">SeqI</A></CODE></B></DD>
<DD>Get the text description of the sequence. This is all the text on the description line
 of a Fasta file after the id.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../javagene/seq/SeqI.html#description()">description</A></CODE> in interface <CODE><A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>The text description.</DL>
</DD>
</DL>
<HR>

<A NAME="id()"><!-- --></A><H3>
id</H3>
<PRE>
public java.lang.String <B>id</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../javagene/seq/SeqI.html#id()">SeqI</A></CODE></B></DD>
<DD>Get the (single-word) identifier of the sequence. This is the first identifier
 in a Fasta file.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../javagene/seq/SeqI.html#id()">id</A></CODE> in interface <CODE><A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>The single-word identifier.</DL>
</DD>
</DL>
<HR>

<A NAME="mapFromBase(javagene.seq.Location)"><!-- --></A><H3>
mapFromBase</H3>
<PRE>
public <A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A> <B>mapFromBase</B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;baseLocation)</PRE>
<DL>
<DD>Map a location on the base sequence to the equivalent
 genomic location. The returned genomic location is 
 always on the same strand as the location bounds.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>baseLocation</CODE> - The location on the base string.
<DT><B>Returns:</B><DD>A location within the location bounds.
<DT><B>Throws:</B>
<DD><CODE>java.lang.IndexOutOfBoundsException</CODE> - The specified baseLocation could not
 be mapped within the SeqStranded object's location bounds.</DL>
</DD>
</DL>
<HR>

<A NAME="mapToBase(javagene.seq.Location)"><!-- --></A><H3>
mapToBase</H3>
<PRE>
public <A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A> <B>mapToBase</B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;genomicLocation)</PRE>
<DL>
<DD>Map a genomic location to the equivalent location
 on the base sequence. If the specified location
 is on the opposite strand, this method first converts
 the specified location to the other strand, and then
 maps the result to the base sequence.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>genomicLocation</CODE> - The location to map. Must be within the location bounds,
 though both strands are allowed.
<DT><B>Returns:</B><DD>The equivalent location on the base sequence.
<DT><B>Throws:</B>
<DD><CODE>java.lang.IndexOutOfBoundsException</CODE> - The genomicLocation did not map
 within the base sequence bounds.</DL>
</DD>
</DL>
<HR>

<A NAME="subseq(javagene.seq.Location)"><!-- --></A><H3>
subseq</H3>
<PRE>
public <A HREF="../../javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A> <B>subseq</B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;location)</PRE>
<DL>
<DD>Create a new sequence containing only the portion specified by the location parameter.
 Shares the underlying data (does not make a copy).

 The specified location must be contained by the bounding location,
 and on the same strand. Consider using the toString( Location ) method
 if the location might be on either strand.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../javagene/seq/SeqI.html#subseq(javagene.seq.Location)">subseq</A></CODE> in interface <CODE><A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>location</CODE> - The location of the subsequence.
<DT><B>Returns:</B><DD>The subsequence.
<DT><B>Throws:</B>
<DD><CODE>java.lang.IndexOutOfBoundsException</CODE> - The bounding location did not
 contain the specified location.</DL>
</DD>
</DL>
<HR>

<A NAME="subseq(javagene.seq.Location, java.lang.String, java.lang.String)"><!-- --></A><H3>
subseq</H3>
<PRE>
public <A HREF="../../javagene/seq/SeqStranded.html" title="class in javagene.seq">SeqStranded</A> <B>subseq</B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;location,
                          java.lang.String&nbsp;id,
                          java.lang.String&nbsp;description)</PRE>
<DL>
<DD>Create a new sequence containing only the portion specified by the location parameter,
 with the specified id and description.
 Shares the underlying data (does not make a copy).

 The specified location must be contained by the bounding location,
 and on the same strand. Consider using the toString( Location ) method
 if the location might be on either strand.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../javagene/seq/SeqI.html#subseq(javagene.seq.Location, java.lang.String, java.lang.String)">subseq</A></CODE> in interface <CODE><A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>location</CODE> - The location of the subsequence.<DD><CODE>id</CODE> - The new single-word id.<DD><CODE>description</CODE> - The new description line.
<DT><B>Returns:</B><DD>The subsequence.
<DT><B>Throws:</B>
<DD><CODE>java.lang.IndexOutOfBoundsException</CODE> - The bounding location did not
 contain the specified location.</DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD>Get the IUPAC character string representation of this entire sequence,
 as defined by the bounding location.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../javagene/seq/SeqI.html#toString()">toString</A></CODE> in interface <CODE><A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></CODE><DT><B>Overrides:</B><DD><CODE>toString</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>The text string.</DL>
</DD>
</DL>
<HR>

<A NAME="toString(javagene.seq.Location)"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>(<A HREF="../../javagene/seq/Location.html" title="class in javagene.seq">Location</A>&nbsp;location)</PRE>
<DL>
<DD>Get the IUPAC character string representation of the subsequence specified by 
 the location parameter. Returns the reverse complement if the specified location
 and bounding locations are on opposite strands. Uses the sequence type (Rna or Dna) 
 to choose whether T or U should be used when complementing (if necessary). 
 The specified location must be 
 within the bounding location.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../javagene/seq/SeqI.html#toString(javagene.seq.Location)">toString</A></CODE> in interface <CODE><A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq">SeqI</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>location</CODE> - The portion of the sequence to get.
<DT><B>Returns:</B><DD>The string.
<DT><B>Throws:</B>
<DD><CODE>java.lang.IndexOutOfBoundsException</CODE> - The location parameter was not within the bounding location
 of the sequence.</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../javagene/seq/SeqI.html" title="interface in javagene.seq"><B>PREV CLASS</B></A>&nbsp;
&nbsp;NEXT CLASS</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?javagene/seq/SeqStranded.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SeqStranded.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
